File: bst.cpp
#include<bits/stdc++.h>
using namespace std;
class Node{
    public:
    int data;
    Node *left = NULL;
    Node *right = NULL;
};
Node *goleft(Node *root){
    while(root->left){
        root = root->left;
    }
    return root;
}

Node* delet(Node *root,int val){
    if(root==NULL){
        return root;
    }
    else if(val<root->data){
        root->left = delet(root->left,val);
    }
    else if(val>root->data){
        root->right = delet(root->right,val);
    }
    else{
        if (root->left == NULL) 
        { 
            Node *temp = root->right; 
            free(root); 
            return temp; 
        } 
        else if (root->right == NULL) 
        { 
            Node *temp = root->left; 
            free(root); 
            return temp; 
        } 
        Node *temp = goleft(root->right);
        root->data = temp->data; 
        root->right = delet(root->right, temp->data); 
    }
    return root;
}

Node *insert(int val,Node *root){
    if(root==NULL){
        Node *temp = new Node();
        temp->data = val;
        root = temp;
        return root;
    }
    if(val < root->data){
        root->left = insert(val,root->left);
    }
    else{
        root->right = insert(val,root->right);
    }
    return root;
}

void display(Node *root){
    if(root==NULL){
        return;
    }
    else{
        display(root->left);
        cout<<root->data<<" ";
        display(root->right);
    }
}

void pushall(int arr[],int n,Node *root){
    for(int i=0;i<n;i++){
        if(root==NULL){
            Node *temp = new Node();
            temp->data = arr[i];
            root = temp;
        }
        else{
            root = insert(arr[i],root);
        }
    }

    display(root);
    cout<<endl;
    root = delet(root,34);
    display(root);
    root = delet(root,67);
    display(root);
}

int main(){
    int n;
    cin>>n;
    int arr[n];
    for(int i=0;i<n;i++){
        cin>>arr[i];
    }
    Node *root = NULL;
    pushall(arr,n,root);
}

File: BST.java
import java.util.*;
import java.io.*;
import java.lang.*;

class Node{
    private int data;
    public Node left = null;
    public Node right = null;

    Node(int d){
        this.data = d;
    }

    public int getData(){
        return this.data;
    }

    public void setData(int d){
        this.data = d;
    }
}

public class BST{

    public static Node insertion(int val, Node root){
        if(root == null){
            return new Node(val);
        }
        
        if(val < root.getData()){
            root.left = insertion(val,root.left);
        }
        else{
            root.right = insertion(val,root.right);
        }
        return root;
    }

    public static Node delete(int val, Node root){
        if(root == null){
            return root;
        }

        if(root.getData() < val){
            root.right = delete(val,root.right);
        }
        else if(root.getData() > val){
            root.left = delete(val, root.left);
        }
        else{
            if(root.left == null){
                return root.right;
            }
            else if(root.right == null){
                return root.left;
            }
            int min = inorderSuccessor(root.right);
            root.setData(min);
            root.right = delete(min,root.right); 
            return root;
        }
        return root;
    }

    public static int inorderSuccessor(Node root){
        int v = root.getData();
        while(root.left!=null){
            root = root.left;
            v = root.getData();
        }
        return v;
    }

    public static void display(Node root){
        if(root == null){
            return;
        }
        else{
            display(root.left);
            System.out.print(root.getData() + " ");
            display(root.right);
        }

    }

    public static void main(String[] args){
        try{
            BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
            System.out.println("Enter size:");
            int s = Integer.parseInt(br.readLine());
            Node root = null;
            for(int i=0;i<s;i++){
                int val = Integer.parseInt(br.readLine());
                root = insertion(val,root);
                display(root);
                System.out.println("---------------------------");
            }

            System.out.println("Enter no of items to delete:");
            int it = Integer.parseInt(br.readLine());
            for(int i=0;i < it;i++){
                int valtodelete = Integer.parseInt(br.readLine());
                root = delete(valtodelete, root);
                display(root);
                System.out.println("---------------------------");
            }

        }
        catch(Exception e){
            System.out.println(e.getMessage());
        }
    }
}